## OpenAI Embedding

In [3]:
import os 
from dotenv import load_dotenv
load_dotenv()

True

In [4]:
os.environ["OPENAI_API_KEY"]=os.getenv("OPENAI_API_KEY")

In [5]:
from langchain_openai import OpenAIEmbeddings
embeddings=OpenAIEmbeddings(model="text-embedding-3-large")
embeddings

OpenAIEmbeddings(client=<openai.resources.embeddings.Embeddings object at 0x116cae690>, async_client=<openai.resources.embeddings.AsyncEmbeddings object at 0x116caf770>, model='text-embedding-3-large', dimensions=None, deployment='text-embedding-ada-002', openai_api_version=None, openai_api_base=None, openai_api_type=None, openai_proxy=None, embedding_ctx_length=8191, openai_api_key=SecretStr('**********'), openai_organization=None, allowed_special=None, disallowed_special=None, chunk_size=1000, max_retries=2, request_timeout=None, headers=None, tiktoken_enabled=True, tiktoken_model_name=None, show_progress_bar=False, model_kwargs={}, skip_empty=False, default_headers=None, default_query=None, retry_min_seconds=4, retry_max_seconds=20, http_client=None, http_async_client=None, check_embedding_ctx_length=True)

In [6]:
text="This is langchain and openAI embedding tutorial"
query_result=embeddings.embed_query(text)
query_result

[0.00204049632884562,
 0.020377667620778084,
 -0.022670665755867958,
 -0.017743447795510292,
 0.024717986583709717,
 -0.0400455966591835,
 0.014481382444500923,
 0.028416812419891357,
 -0.01465881709009409,
 0.01235216949135065,
 -0.02030942216515541,
 0.01756601221859455,
 -0.022520529106259346,
 -0.01669248938560486,
 0.01577802002429962,
 -0.007984551601111889,
 0.005053470376878977,
 0.032456859946250916,
 0.006575311999768019,
 -0.01191540714353323,
 -0.0024909069761633873,
 0.006653792690485716,
 -0.04083722457289696,
 -0.005995237734168768,
 0.02013198845088482,
 -0.020241178572177887,
 0.026069218292832375,
 0.021510517224669456,
 -0.0040025124326348305,
 0.024335820227861404,
 0.02018658258020878,
 -0.028471408411860466,
 -0.0008560360292904079,
 -0.02550961822271347,
 -0.013662454672157764,
 -0.01262514479458332,
 0.059290412813425064,
 0.05443143844604492,
 -0.03756151348352432,
 0.013212043792009354,
 0.01560058444738388,
 0.025714349001646042,
 -0.02833492122590542,
 -0.02

In [7]:
len(query_result)

3072

In [8]:
embeding_1024=OpenAIEmbeddings(model="text-embedding-3-large",dimensions=1024)

In [9]:
embeding_1024

OpenAIEmbeddings(client=<openai.resources.embeddings.Embeddings object at 0x1305c6b40>, async_client=<openai.resources.embeddings.AsyncEmbeddings object at 0x114d606e0>, model='text-embedding-3-large', dimensions=1024, deployment='text-embedding-ada-002', openai_api_version=None, openai_api_base=None, openai_api_type=None, openai_proxy=None, embedding_ctx_length=8191, openai_api_key=SecretStr('**********'), openai_organization=None, allowed_special=None, disallowed_special=None, chunk_size=1000, max_retries=2, request_timeout=None, headers=None, tiktoken_enabled=True, tiktoken_model_name=None, show_progress_bar=False, model_kwargs={}, skip_empty=False, default_headers=None, default_query=None, retry_min_seconds=4, retry_max_seconds=20, http_client=None, http_async_client=None, check_embedding_ctx_length=True)

In [10]:
text="This is langchain and openAI embedding tutorial"
query_result=embeding_1024.embed_query(text)
query_result

[0.002796480432152748,
 0.02804463542997837,
 -0.03125671297311783,
 -0.02443809248507023,
 0.03399918973445892,
 -0.05507493019104004,
 0.019911127164959908,
 0.039070893079042435,
 -0.02013653703033924,
 0.017008986324071884,
 -0.027950715273618698,
 0.02423146739602089,
 -0.030993737280368805,
 -0.022954149171710014,
 0.02167683094739914,
 -0.011016864329576492,
 0.006950110662728548,
 0.04470611736178398,
 0.009006967768073082,
 -0.016426680609583855,
 -0.003376438980922103,
 0.009119671769440174,
 -0.05616440623998642,
 -0.008297869004309177,
 0.027687737718224525,
 -0.02787557803094387,
 0.03584003075957298,
 0.029603714123368263,
 -0.005550696514546871,
 0.03345445170998573,
 0.027819227427244186,
 -0.03922116383910179,
 -0.0011346760438755155,
 -0.03503231331706047,
 -0.018812259659171104,
 -0.017403453588485718,
 0.08159805834293365,
 0.07494848966598511,
 -0.05173136293888092,
 0.01820177584886551,
 0.0214702058583498,
 0.03540799766778946,
 -0.038995757699012756,
 -0.0290589

In [11]:
len(query_result)

1024

# Chroma Vector DB

In [12]:
from langchain_community.document_loaders import TextLoader

loader=TextLoader("data/loreal_shareholder_2022.txt")
loader

In [13]:
docs=loader.load()
docs

[Document(metadata={'source': 'data/loreal_shareholder_2022.txt'}, page_content='“Dear Shareholders,\nL’Oréal continues on the path to success with an ever-stronger ambition, while acting with the sense of responsibility of a global leader. Dual financial and social excellence will always be at the heart of our business model.\nWe have set ourselves the ultimate goal of creating value that benefits everyone.\nWe create value for you, our shareholders.\xa0The resilience and outperformance of your Company are the perfect demonstration of its robust, virtuous and value creating business model. The quality of our results puts us in a position to offer a dividend of €6 per share, representing a significant increase of +25%. And the preferential dividend with a 10% loyalty bonus(1), at €6.60, is recognition of your long-term loyalty.\nI also know that you attach just as much importance to the quality of our relationship with you, our shareholders. I am delighted to welcome the more than 30,0

In [14]:
from langchain_text_splitters import RecursiveCharacterTextSplitter

text_splitter=RecursiveCharacterTextSplitter(chunk_size=500,chunk_overlap=50)
final_documents=text_splitter.split_documents(docs)
final_documents

[Document(metadata={'source': 'data/loreal_shareholder_2022.txt'}, page_content='“Dear Shareholders,\nL’Oréal continues on the path to success with an ever-stronger ambition, while acting with the sense of responsibility of a global leader. Dual financial and social excellence will always be at the heart of our business model.\nWe have set ourselves the ultimate goal of creating value that benefits everyone.'),
 Document(metadata={'source': 'data/loreal_shareholder_2022.txt'}, page_content='We create value for you, our shareholders.\xa0The resilience and outperformance of your Company are the perfect demonstration of its robust, virtuous and value creating business model. The quality of our results puts us in a position to offer a dividend of €6 per share, representing a significant increase of +25%. And the preferential dividend with a 10% loyalty bonus(1), at €6.60, is recognition of your long-term loyalty.'),
 Document(metadata={'source': 'data/loreal_shareholder_2022.txt'}, page_co

In [15]:
from langchain_community.vectorstores import Chroma

db=Chroma.from_documents(final_documents,embeding_1024)

ImportError: Could not import chromadb python package. Please install it with `pip install chromadb`.

In [16]:
## Retreive the result form query vectorstore db

query="Loreal revenue"
retrived_result=db.similarity_search(query)
print(retrived_result)

[Document(metadata={'source': 'data/loreal_shareholder_2022.txt'}, page_content='We create value for our ecosystem.\xa0L’Oréal is a global company that thrives on its diversity. But our heart beats in France and always will. We have nearly\xa015,000 employees\xa0in the country. Our French plants handle more than 26% of our worldwide production, even though France represents 7% of our sales. We have made France a land of the Group’s scientific and technological excellence, with\xa02,700 researchers\xa0and nearly 70% of our R&D investments.'), Document(metadata={'source': 'data/loreal_shareholder_2022.txt'}, page_content='It was only natural that we share L’Oréal’s performance with them. In 2022, we launched a third employee share ownership plan. More than a third of our employees around the world and two-thirds of employees in France are Group shareholders. That is outstanding proof of their attachment to L’Oréal. And it is yet another means of aligning their interests with yours.'), Do

Example:
If you embed "ChatGPT" using:

OpenAI Embedding (text-embedding-ada-002)
→ [0.021, 0.332, ..., 0.019] (1536D)

Ollama Embedding (e.g., nomic-embed-text)
→ [0.0012, -0.075, ..., 0.111] (768D or 1024D)

They won’t be equal. But each vector will still meaningfully represent the semantics of "ChatGPT" in its own latent space.


LLMs don’t "understand" embeddings from other LLMs natively.
Each model operates in its own embedding space. If you feed vectors from Model A into Model B, Model B won’t know what they mean unless they've been aligned somehow.


